Add CUDA cache cleaning flag to pytorch backend

zeruniverse opened the pull request:
Implemented a flag similar to `memory_arena_shrinkage` in ONNXRuntime backend. With `ENABLE_CACHE_CLEANING` set to `true`, CUDA cache gets cleared after each model execution. This will enable us to serve more models in one GPU. 

I have tested on local GPU: with `enable_cache_cleaning=false`, the model takes additional 2Gi GRAM after running inference with batch size 30. After that, this 2Gi never gets shrunk. With `enable_cache_cleaning=true`, the additional GRAM gets shrunk every time after model execution.

manhtd98 commented:
why we not merge it? i have the same problem

tanmayv25 commented:
This might have fell through the cracks. Thanks for bringing it up.

tanmayv25 commented on src/libtorch.cc:
```diff
@@ -445,6 +469,9 @@ class ModelInstanceState : public BackendModelInstance {
   void ProcessRequests(
       TRITONBACKEND_Request** requests, const uint32_t request_count);
 
+  // Clear cache
```
Clear CUDA cache

tanmayv25 commented on src/libtorch.cc:
```diff
@@ -1551,6 +1583,8 @@ TRITONBACKEND_ModelInstanceExecute(
   // specific request.
   instance_state->ProcessRequests(requests, request_count);
 
+  if(model_state->EnabledCacheCleaning()) instance_state->ClearCache();
```
enclose the call  within braces.

tanmayv25 commented on README.md:
````diff
@@ -179,6 +179,20 @@ key: "ENABLE_WEIGHT_SHARING"
 ```
 
+* `ENABLE_CACHE_CLEANING`: Boolean flag to enable CUDA cache cleaning after each model execution.
+If not specified, cache cleaning is disabled. This flag has no effect if model is on CPU.
````
You should also talk about how it would penalize the performance.

zeruniverse commented:
@tanmayv25 do you have a CLA version for individual contributor? This CLA is for corporation.

tanmayv25 commented:
@zeruniverse  Just provide your name instance of company name in the form. We use the same form for company and individual.

zeruniverse commented:
@tanmayv25 

- [x] CLA sent to triton-cla at nvidia dot com

- [x] addressed review comments

- [ ] perf data (do we need it? I have no hardware to do a thorough perf test)

tanmayv25 commented:
Thanks @zeruniverse! I will add some sanity testing for this on our CI and merge the changes.
